arrow/formatter.py

```diff
@@ -22,19 +31,24 @@ class DateTimeFormatter:
     # an atomic group. For more info on atomic groups and how to they are
     # emulated in Python's re library, see https://stackoverflow.com/a/13577411/2701578
 
-    _FORMAT_RE = re.compile(
+    _FORMAT_RE: Final[Pattern[str]] = re.compile(
         r"(\[(?:(?=(?P<literal>[^]]))(?P=literal))*\]|YYY?Y?|MM?M?M?|Do|DD?D?D?|d?dd?d?|HH?|hh?|mm?|ss?|SS?S?S?S?S?|ZZ?Z?|a|A|X|x|W)"
     )
 
-    def __init__(self, locale="en_us"):
+    locale: locales.Locale
+
+    def __init__(self, locale: str = "en_us") -> None:
 
         self.locale = locales.get_locale(locale)
 
-    def format(cls, dt, fmt):
+    def format(cls, dt: datetime, fmt: str) -> str:
 
-        return cls._FORMAT_RE.sub(lambda m: cls._format_token(dt, m.group(0)), fmt)
+        # FIXME: _format_token() is nullable
```

Comment:
@jadchaar Nop. `DateTimeFormatter._format_token()` still can returns `None`. (if `token` did not match with all conditions) But because `DateTimeFormatter._format_token()` is internal purpose, we can override return type as `str` and remove cast and FIXME

Reply:
@jadchaar And plus, we have to make sure that `dt` is not naive. `dt.tzname()` in `DateTimeFormatter._format_token()` can return `None` if `dt` is naive.

Reply:
@isac322 is this FIXME good to be removed?

Reply:
@jadchaar I can't because return type of `_format_token()` is still `Optional[str]` as explained [comment below](https://github.com/arrow-py/arrow/pull/883#discussion_r575953514)